### PR TITLE
[ValueTracking] Implement `computeKnownFPClass` for `llvm.vector.reduce.{fmin,fmax,fmaximum,fminimum}`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[ValueTracking] Add tests for `computeKnownFPClass` of `llvm.vector.reduce.{fmin,fmax,fmaximum,fminimum}`; NFC**
- **[ValueTracking] Implement `computeKnownFPClass` for `llvm.vector.reduce.{fmin,fmax,fmaximum,fminimum}`**

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -5004,6 +5004,15 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
 
       break;
     }
+      // reduce min/max will choose an element from one of the vector elements,
+      // so we can infer and class information that is common to all elements.
+    case Intrinsic::vector_reduce_fmax:
+    case Intrinsic::vector_reduce_fmin:
+    case Intrinsic::vector_reduce_fmaximum:
+    case Intrinsic::vector_reduce_fminimum:
+      computeKnownFPClass(II->getArgOperand(0), Known, InterestedClasses,
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:5013]: You cannot keep the signbit information if we don't know it is never nan.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:5013]: I need to add proofs for all of the cases. I will do so before proceeding w/ this.

File: llvm/test/Transforms/InstSimplify/known-never-infinity.ll [hunk 2]
```diff
@@ -1109,6 +1109,106 @@ define float @fcmp_ult_neginf_implies_class_assert(float %arg) {
...
+define i1 @isKnownNeverInfinity_vector_reduce_maximum_fail(<4 x double> %x) {
+; CHECK-LABEL: define i1 @isKnownNeverInfinity_vector_reduce_maximum_fail
+; CHECK-SAME: (<4 x double> [[X:%.*]]) {
+; CHECK-NEXT:    [[NINF_X:%.*]] = fadd <4 x double> [[X]], <double 1.000000e+00, double 1.000000e+00, double 1.000000e+00, double 1.000000e+00>
+; CHECK-NEXT:    [[OP:%.*]] = call double @llvm.vector.reduce.fmaximum.v4f64(<4 x double> [[NINF_X]])
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp une double [[OP]], 0x7FF0000000000000
+; CHECK-NEXT:    ret i1 [[CMP]]
+;
+  %ninf.x = fadd <4 x double> %x, <double 1.0, double 1.0, double 1.0, double 1.0>
+  %op = call double @llvm.vector.reduce.fmaximum(<4 x double> %ninf.x)
```
Comment (arsenm) [inline llvm/test/Transforms/InstSimplify/known-never-infinity.ll:1132]: Missing the .v4f64

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Analysis/ValueTracking.cpp:5013]: For the fmin/fmax case you only need to know one element is non-nan

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:5013]: Plan now isn't to do per-ele, although I can leave that as a todo.

Also looks like alive2 doesn't support fp reduce intrins.

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:5013]: Fixed + tests.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -5007,6 +5007,19 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
         Known.knownNot(fcNegZero);
 
       break;
+    }
+      // reduce min/max will choose an element from one of the vector elements,
+      // so we can infer and class information that is common to all elements.
```
Comment (arsenm) [inline llvm/lib/Analysis/ValueTracking.cpp:5012]: Move comment inside the cases?

File: llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+; CHECK-NEXT:    [[OP:%.*]] = call double @llvm.vector.reduce.fmin.v4f64(<4 x double> [[X_ABS]])
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp oge double [[OP]], 0.000000e+00
+; CHECK-NEXT:    ret i1 [[CMP]]
+;
+  %x.abs = call <4 x double> @llvm.fabs.v4f64(<4 x double> %x)
+  %op = call double @llvm.vector.reduce.fmin.v4f64(<4 x double> %x.abs)
+  %cmp = fcmp oge double %op, 0.0
+  ret i1 %cmp
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll:105]: Should test the case where the nnan flag on the call itself allows preserving the signbit

**Review by arsenm — APPROVED** [review]

File: llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+; RUN: opt < %s -S -passes=instcombine | FileCheck %s
+
+define i1 @vector_reduce_maximum_signbit(<4 x double> nofpclass(nan nzero) %x) {
+; CHECK-LABEL: define i1 @vector_reduce_maximum_signbit
+; CHECK-SAME: (<4 x double> nofpclass(nan nzero) [[X:%.*]]) {
+; CHECK-NEXT:    ret i1 true
+;
+  %x.abs = call <4 x double> @llvm.fabs.v4f64(<4 x double> %x)
+  %op = call double @llvm.vector.reduce.fmaximum.v4f64(<4 x double> %x.abs)
+  %cmp = fcmp oge double %op, 0.0
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll:11]: Why?

[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll:11]: It is the canonical form of signbit idiom: https://godbolt.org/z/zMnPPq6o3. I prefer to use the canonical form, although in cases with nnan/nsz both forms are fine.

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll:11]: That is more instructions, so it probably shouldn't be the canonical form

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll:11]: Although I suppose it works even in the nan case

[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll:11]: `bitcast + icmp` doesn't work. Needs for `knownFPClass` -> `KnownBits` patch. My preference would be to just get this in, but I can wait if there are strong feelings.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll (hunk 5 above)
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/known-fpclass-reduce-signbit.ll:11]: For the purposes of this patch, I think it should just stay the fcmp. It's fewer operations and has the same effect for the test

**Pull request closed without merging**